### ISSUE TITLE
Add support for events.

### ISSUE THREAD

**Issue by ixti** [httprb/http.rb#209]
Proposed API:

``` ruby
# Add default (global) `:connect` event listener 
HTTP.on(:connect) { ... }

# Add `:request` event listener for particular client instance
client = HTTP.auth("Bearer 123")
client.on(:request) { ... }
```

With proper architecture, IMO it will become possible to extract timeouts out of connection/client completely.

**Comment by zanker** [comment]
Timeouts requires you have direct access to socket write and reads. You can't do that with an API.

I'm not sure I see an advantage to this. Celluloid already lets you do the async bits. Seems like unneeded abstraction.

**Comment by tarcieri** [comment]
@zanker this kind of API can be useful for people writing things like testing frameworks who want to instrument clients for various purposes. Right now they're using monkeypatches.

See this discussion:

https://groups.google.com/forum/#!topic/httprb/ISln457aPUc

**Comment by ixti** [comment]
What comes to VCR support (from ML) actually that can't be done without monkey patching anyway :D And it's already done (webmock backend for VCR). I was mostly thinking about kind of a way to "monitor" progress. Like log requests and timings, probably sending "halt" to client on some events...

**Comment by tarcieri** [comment]
It'd be nice to have something like `Reel::Spy` in http.rb:

https://github.com/celluloid/reel/wiki/Spy

A system like this would make it easy to implement.

**Comment by ixti** [comment]
Proposed API:

``` ruby
module HTTP
  def self.spy(spy = Spy, &block)
    case
    when spy.is_a?(Spy)
      branch(:spy => spy)
    when spy.is_a?(Class) && spy.ancestors.include?(Spy)
      branch :spy => spy.new(&block)
    else
      fail "Instance or subclass of Spy expected"
    end
  end
end

client_with_spy = HTTP.spy do
  on(:connect) { |client| ... }
  on(:request) { |client| ... }
end
```

**Comment by pezra** [comment]
Would either of the proposed APsI allow modification of the request flow. Ie, could an `on(:request)` bypass the actual HTTP request in favor of a cached response instead? Allowing a "spy" to do that seems a little unintuitive but maybe not out of the question.

**Comment by tarcieri** [comment]
@pezra it could. It doesn't have to be called "spy"

**Comment by ixti** [comment]
@tarcieri if we are talking about events, then I don't see how it can modify workflow. I mean I would like to avoid (and I know you too) middleware alike architecture.

**Comment by Senjai** [comment]
RE: logging, I probably wouldn't complicate it too much. I think hooks as extension points would be great, but they're not required to make logging better. Couldn't we just add a debug option on configuration and a logger class? If we choose to add hooks for extension, the logger could be registered as a default hook if the option is true.

Until that time, I think these could be two separate issues. Any opinions?

**Comment by Senjai** [comment]
@pezra I think hooks should be different from composition. Hooks should allow you to execute code when a thing occurs. But it should not block other code from running. 

Instead it'd be nice to have a connection class, or request class or whatever that can be substituted in favor of a dummy request object via composition.

**Comment by pezra** [comment]
@ixti, I agree that allowing an event handler to modify the flow send awkward.

@senjai, an option specify the connection class to use might be workable. Wild such a PR be accepted?

**Comment by tarcieri** [comment]
@Senjai we already have `HTTP::Connection` and `HTTP::Request` classes

@pezra if it's not too invasive and doesn't involve weird delegators/proxy objects
